File: src/main/java/com/indiepost/model/Tag.java
package com.indiepost.model;

import javax.persistence.*;
import javax.validation.constraints.Size;
import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

/**
 * Created by jake on 7/25/16.
 */

@Entity
@Table(name = "Tags")
public class Tag implements Serializable {

    private static final long serialVersionUID = 4776718128460861941L;

    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    private Long id;

    @Size(min = 1, max = 50)
    @Column(nullable = false, unique = true)
    private String name;

    @OneToMany(
            mappedBy = "tag",
            cascade = CascadeType.ALL,
            orphanRemoval = true
    )
    private List<PostTag> postTags = new ArrayList<>();

    public Tag() {
    }

    public Tag(String name) {
        this.name = name;
    }

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public List<PostTag> getPostTags() {
        return postTags;
    }

    public void setPostTags(List<PostTag> postTags) {
        this.postTags = postTags;
    }

    public List<Post> getPosts() {
        return postTags.stream()
                .map(postTag -> postTag.getPost())
                .collect(Collectors.toList());
    }
}

File: src/main/kotlin/com/indiepost/model/Tag.kt
package com.indiepost.model

import javax.persistence.*
import javax.validation.constraints.Size

/**
 * Created by jake on 7/25/16.
 */

@Entity
@Table(name = "Tags")
data class Tag(
        @Id
        @GeneratedValue(strategy = GenerationType.AUTO)
        var id: Long? = null,

        @Size(min = 1, max = 50)
        @Column(nullable = false, unique = true)
        var name: String? = null,

        @OneToMany(mappedBy = "tag", cascade = [CascadeType.ALL], orphanRemoval = true)
        var postTags: MutableList<PostTag> = ArrayList()
)

